src/Visitor.ts

```diff
@@ -0,0 +1,234 @@
+import * as ts from 'typescript'
+
+import { Counter } from './Counter'
+import { Descriptor } from './Descriptor'
+import { Input } from './Input'
+import * as lsif from './lsif'
+import { LsifSymbol } from './LsifSymbol'
+import { lsif_typed } from './main'
+import { Packages } from './Packages'
+import { Range } from './Range'
+
+export class Visitor {
+  private localCounter = new Counter()
+  private propertyCounters: Map<string, Counter> = new Map()
+  private localSymbolTable: Map<ts.Node, LsifSymbol> = new Map()
+  constructor(
+    public readonly checker: ts.TypeChecker,
+    public readonly input: Input,
+    public readonly document: lsif.lib.codeintel.lsif_typed.Document,
+    public readonly globalSymbolTable: Map<ts.Node, LsifSymbol>,
+    public readonly packages: Packages,
+    public readonly sourceFile: ts.SourceFile
+  ) {}
+  public index(): void {
+    this.visit(this.sourceFile)
+  }
+  private visit(node: ts.Node): void {
+    if (ts.isIdentifier(node)) {
+      const sym = this.checker.getSymbolAtLocation(node)
+      if (sym) {
+        this.visitIdentifier(node, sym)
+      }
+    }
+    ts.forEachChild(node, node => this.visit(node))
+  }
+
+  private visitIdentifier(node: ts.Identifier, sym: ts.Symbol): void {
+    const range = Range.fromNode(node).toLsif()
+    let role = 0
+    const isDefinition = this.declarationName(node.parent) === node
+    if (isDefinition) {
+      role |= lsif_typed.SymbolRole.Definition
+    }
+    for (const declaration of sym?.declarations || []) {
+      const lsifSymbol = this.lsifSymbol(declaration)
+      this.document.occurrences.push(
+        new lsif.lib.codeintel.lsif_typed.Occurrence({
+          range,
+          symbol: lsifSymbol.value,
+          symbol_roles: role,
+        })
+      )
```

Comment:
Is it possible for `lsifSymbol` here to be empty? If not, would be good to some check + assert or check + emit warning. If it's possible, does it make sense to still add that occurrence? (Similarly for the other place where there's a `this.lsifSymbol` call followed by an `occurrences.push` call.)